### PR TITLE
Update tilemap physics' World2D on reparenting

### LINKED ISSUE #83291: Reparenting a CharacterBody2D/TileMap duo to or from a SubViewport disables their collisions
### Godot version

4.2.beta4

### System information

Godot v4.2.beta4 - Windows 10.0.19045 - GLES3 (Compatibility) - Intel(R) UHD Graphics 620 (Intel Corporation; 31.0.101.2125) - Intel(R) Core(TM) i5-8250U CPU @ 1.60GHz (8 Threads)

### Issue description

Pretty much what the title says. Reparenting a common ancestor of a CharacterBody2D and a TileMap to or from a SubViewport will cause their collisions with each other not to work. Interestingly enough, changing the `collision_animatable` field of the TileMap - which I guess flushes its collision or something - after reparenting, fixes this issue.

To be sure, I've checked reparenting to a regular Node2D, and also to a plain Node - it seems that this bug is specific to reparenting to a SubViewport. I've also not had this issue with collision objects other than a TileMap (static collision objects for instance do not experience this bug).

### Steps to reproduce

Attached below is a minimal reproduction project, with a CharacterBody2D moving up and down in a loop, and a tilemap containing two bars that block the CharacterBody2D's motion.

Pressing the down key reparents them to a SubViewport node, which is displayed on screen via a Sprite2D with a ViewportTexture (intentionally slightly smaller so you can distinguish the difference). If it's pressed again, the two are reparented back to their original node. You will notice that after pressing the down key, the CharacterBody2D will pass through the tilemap.

Pressing the up key will toggle the tilemap's `collision_animatable` flag, fixing the bug until the down key is pressed again.

### Minimal reproduction project

[tilemap_reparent_mrp.zip](https://github.com/godotengine/godot/files/12898684/tilemap_reparent_mrp.zip)

### PR DESCRIPTION
**Pull request by groud**
Supersedes https://github.com/godotengine/godot/pull/84757
Fixes https://github.com/godotengine/godot/issues/83291

This is the same fix as in #84757, with the changes I requested. Made this superseding PR to have this included in RC1

### CODE REVIEW

File: scene/2d/tile_map.cpp [hunk 1]
```diff
@@ -714,7 +718,22 @@ void TileMapLayer::_physics_notify_tilemap_change(TileMapLayer::DirtyFlags p_wha
...
 						Transform2D xform(0, tile_map_node->map_to_local(bodies_coords[body]));
 						xform = gl_transform * xform;
-						PhysicsServer2D::get_singleton()->body_set_state(body, PhysicsServer2D::BODY_STATE_TRANSFORM, xform);
+						ps->body_set_state(body, PhysicsServer2D::BODY_STATE_TRANSFORM, xform);
+					}
+				}
+			}
+		}
+	} else if (p_what == DIRTY_FLAGS_TILE_MAP_IN_TREE) {
+		// Changes in the tree may cause the space to change (e.g. when reparenting to a SubViewport)
```
Comment (akien-mga) [inline scene/2d/tile_map.cpp:727]:
```suggestion
		// Changes in the tree may cause the space to change (e.g. when reparenting to a SubViewport).
```

**Review by akien-mga — APPROVED** [review]

File: scene/2d/tile_map.cpp [hunk 2]
```diff
@@ -686,26 +686,30 @@ void TileMapLayer::_physics_update() {
 
 void TileMapLayer::_physics_notify_tilemap_change(TileMapLayer::DirtyFlags p_what) {
 	Transform2D gl_transform = tile_map_node->get_global_transform();
+	PhysicsServer2D *ps = PhysicsServer2D::get_singleton();
```
Comment (AThousandShips) [inline scene/2d/tile_map.cpp:689]: Don't know that storing the server in a variable for three uses is necessary and makes the code harder to read IMO

Comment (akien-mga) [reply scene/2d/tile_map.cpp:689]: I tend to agree it's not necessary, but it doesn't seem to be a bad change either, so I'm fine merging as is (and mostly in a hurry to get the `master` branch somewhat RC1 ready ;)).

**Comment by akien-mga** [comment]
Thanks @NewDefectus and @groud!

File: scene/2d/tile_map.cpp (hunk 2 above)
Comment (groud) [reply scene/2d/tile_map.cpp:689]: > Don't know that storing the server in a variable for three uses is necessary and makes the code harder to read IMO (leftover from the original PR and was planning to comment the same there)

It's commonly done in the TileMap codebase (using `ps`, `rs` or `ns` for the different servers), so it's more or less consistent with what we already have I believe. I think anyone familiar with the TileMap-related code should be familiar with this, so it's does not hurt readability much I think. But yeah, that's debatable.

**Pull request merged**